Updating added types for `ReadableStreamReadDoneResult`

kraenhansen opened the pull request:
This fixes #1675 by generating:

```typescript
type ReadableStreamReadResult<T> = ReadableStreamReadValueResult<T> | ReadableStreamReadDoneResult;

interface ReadableStreamReadDoneResult {
    done: true;
    value: undefined;
}

interface ReadableStreamReadValueResult<T> {
    done: false;
    value: T;
}
```

saschanaz commented:
Makes sense, bu you need to build and commit the result.

saschanaz commented:
Actually byte streams indeed return the value, btw. Perhaps split the type? https://streams.spec.whatwg.org/#byob-reader-internal-slots

saschanaz commented:
Hi @MattiasBuelens, would you be satisfied by this?

Also it would be good to have some unit tests here.

kraenhansen commented:
I've updated the code to separate the `ReadableStreamReadResult` into two:
- `ReadableStreamDefaultReadResult`
- `ReadableStreamBYOBReadResult`

I've also added `ReadableStreamReadResult` which union both, to keep breaking changes to a minimum.

Which re-use the `ReadableStreamReadValueResult` but have their separate "DoneResult" types:
- `ReadableStreamDefaultReadDoneResult`
- `ReadableStreamBYOBReadDoneResult`

The PR is currently failing tests:

```
Test failed: could not compile 'dom.generated.d.ts':
generated/dom.generated.d.ts(28201,75): error TS2552: Cannot find name 'ReadableStreamBYOBReadDoneResult'. Did you mean 'ReadableStreamBYOBReadResult'?
generated/dom.generated.d.ts(28203,78): error TS2552: Cannot find name 'ReadableStreamDefaultReadDoneResult'. Did you mean 'ReadableStreamDefaultReadResult'?
```

I can't seem to get the specialised `*DoneResult` types into the `dom.generated.d.ts` 🤔

MattiasBuelens commented on inputfiles/addedTypes.jsonc:
```diff
@@ -1303,7 +1320,8 @@
                         "value": {
                             "name": "value",
-                            "overrideType": "T"
+                            "overrideType": "T",
```
`ReadableStreamBYOBReadDoneResult.value` should be `T | undefined`, as [per spec](https://streams.spec.whatwg.org/#ref-for-byob-reader-read%E2%91%A4):
> If the reader is canceled, the promise will be fulfilled with an object of the form `{ value: undefined, done: true }`. In this case, the backing memory region of view is discarded and not returned to the caller.

See also: [Deno](https://github.com/denoland/deno/blob/v1.39.4/ext/web/lib.deno_web.d.ts#L609-L624) and [web-streams-polyfill](https://github.com/MattiasBuelens/web-streams-polyfill/blob/v3.3.2/src/lib/readable-stream/byob-reader.ts#L27-L38).

MattiasBuelens commented on inputfiles/addedTypes.jsonc:
```diff
@@ -1303,7 +1320,8 @@
                             "name": "value",
-                            "overrideType": "T"
+                            "overrideType": "T",
+                            "required": true
```
If we make `value: T | undefined`, should we also make it optional? Or does that not work well together with [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes)?

kraenhansen replied:
I don't think so, the spec is pretty clear that the `value` property would exist: https://streams.spec.whatwg.org/#default-reader-read

kraenhansen replied:
You're right. Reading the spec you've linked to, it seems pretty explicit that the `value` property will be in the object returned from `read`, even from a `ReadableStreamBYOBReader`.

MattiasBuelens commented:
> The PR is currently failing tests:
> 
> ```
> Test failed: could not compile 'dom.generated.d.ts':
> generated/dom.generated.d.ts(28201,75): error TS2552: Cannot find name 'ReadableStreamBYOBReadDoneResult'. Did you mean 'ReadableStreamBYOBReadResult'?
> generated/dom.generated.d.ts(28203,78): error TS2552: Cannot find name 'ReadableStreamDefaultReadDoneResult'. Did you mean 'ReadableStreamDefaultReadResult'?
> ```
> 
> I can't seem to get the specialised `*DoneResult` types into the `dom.generated.d.ts` 🤔

The Streams types have `[Exposed=*]` in their IDL, so I think you need to add `"exposed": "*"` (or something equivalent) to your `addedTypes.jsonc` changes?

kraenhansen commented:
> I can't seem to get the specialised *DoneResult types into the dom.generated.d.ts 🤔

I found the issue 👍
